substrate/client/network/sync/src/engine.rs

```diff
@@ -817,8 +882,11 @@ where
 			Poll::Pending => {},
 		}
 
-		// Drive `ChainSync`.
-		while let Poll::Ready(()) = self.chain_sync.poll(cx) {}
+		// Send outbound requests on `ChanSync`'s behalf.
+		self.send_chain_sync_requests();
+
+		// Poll & process pending responses.
+		let _ = self.poll_pending_responses(cx);
```

Comment:
This should be converted to implement `futures::Stream`. I believe that is the last `poll()` that needs to be dealt with (apart from maybe the timer) after `SyncingEngine` could start using `select!` and we can get rid of the hideous `async fn run() { loop { self.poll() }`

Reply:
@altonen I've introduced `PendingResponses` stream, but there are some issues that have to be resolved:

1. Clearly some tests are needed for `PendingResponses`
2. I'm not happy that removing already resolved but not yet processed pending response is linear in terms of the number of `finished_responses`.
3. On every invocation of `PendingResponses::poll_next()` we poll all `pending_responses`, even though some might be already available as `finished_responses`. This one is confusing me, because on one hand looping and polling responses is not free, but on another they still need to make progress.

Reply:
Obvious solution to no.2 is to keep `finished_responses` in a `HashMap` instead of `VecDeque`. This would introduce some randomness in the finished responses dispatching though.

Reply:
Not relevant anymore with `StreamMap`, resolving.